fairscale/experimental/optim/dynamic_loss_scaler.py

```diff
@@ -0,0 +1,203 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+To prevent underflow or overflow of gradients, Dynamic Loss Scaler is used to
+dynamically scale up and down gradients by scaling the loss.
+"""
+
+
+from collections import defaultdict
+from enum import Enum
+
+
+import torch
+
+
+class OptState(Enum):
+    READY = 0
+    UNSCALED = 1
+    STEPPED = 2
+
+
+def _refresh_per_optimizer_state():
+    return OptState.READY
+
+
+class DynamicLossScaler(object):
+    """An instance ``scaler`` helps perform the steps of gradient scaling
+    conveniently.
+
+    * ``scaler.scale(loss)`` multiplies a given loss by ``scaler``'s current scale factor.
+    * ``scaler.step(optimizer)`` safely unscales gradients and calls ``optimizer.step()``.
+    * ``scaler.update()`` updates ``scaler``'s scale factor.
+    """
+
+    def __init__(
+        self,
+        init_scale=2.0 ** 15,
+        scale_factor=2.0,
+        scale_window=2000,
+        tolerance=0.0,
+        threshold=None,
+        min_loss_scale=1e-4,
+    ):
+        self.loss_scale = init_scale
+        self.scale_factor = scale_factor
+        self.scale_window = scale_window
+        self.tolerance = tolerance
+        self.threshold = threshold
+        self._iter = 0
+        self._last_overflow_iter = -1
+        self._last_rescale_iter = -1
+        self._overflows_since_rescale = 0
+        self.min_loss_scale = min_loss_scale
+        self._per_optimizer_states = defaultdict(_refresh_per_optimizer_state)
+        self._scale = None
+
+    def scale(self, outputs):
+        """
+        Multiplies ('scales') a tensor or list of tensors by the scale factor.
+
+        Returns scaled outputs.
+
+        Args:
+            outputs (Tensor or iterable of Tensors):  Outputs to scale.
+        """
+        return self.loss_scale * outputs
+
+    @torch.no_grad()
+    def _get_gradients_norm(self, params) -> float:
+        grads = []
+        for p in params:
+            if p.grad is None:
+                continue
+            else:
+                grads.append(p.grad.detach())
+
+        if len(grads) == 0:
+            return torch.tensor(0.0)
+
+        if len(grads) == 1:
+            total_norm = torch.norm(grads[0], p=2, dtype=torch.float32)
+        else:
+            total_norm = torch.norm(torch.stack([torch.norm(g, p=2, dtype=torch.float32) for g in grads]))
+        return total_norm.item()
+
+    def _decrease_loss_scale(self):
+        self.loss_scale /= self.scale_factor
+        if self.threshold is not None:
+            self.loss_scale = max(self.loss_scale, self.threshold)
+
+    def _check_overflow(self, grad_norm):
+        # detect inf and nan
```

Comment:
could you elaborate a little in the docstring about the plan here ? it looks like this can raise exceptions, but seeing where it's called it does not look like it's completely the plan, I was wondering whether this was on purpose ? The [torch AMP scaler ](https://github.com/pytorch/pytorch/blob/master/torch/cuda/amp/grad_scaler.py)does not raises too aggressively, so I'm wondering whether there's a philosophical difference that I missed ?

Reply:
I added the line to catch the exception in the step function in the 6th commit. When the exception is raised, it means there is INF or NAN in gradients, the optimizer step will be skipped.